sdks/java/extensions/sorter/src/test/java/org/apache/beam/sdk/extensions/sorter/BufferedExternalSorterTest.java

```diff
@@ -0,0 +1,169 @@
+    InMemorySorter mockInMemorySorter = mock(InMemorySorter.class);
+    Constructor<BufferedExternalSorter> constructor =
+        BufferedExternalSorter.class
+            .getDeclaredConstructor(ExternalSorter.class, InMemorySorter.class);
+    constructor.setAccessible(true);
+    BufferedExternalSorter testSorter =
+        constructor.newInstance(mockExternalSorter, mockInMemorySorter);
+
+    @SuppressWarnings("unchecked")
+    KV<byte[], byte[]>[] kvs =
+        new KV[] {
+          KV.of(new byte[] {0}, new byte[] {}),
+          KV.of(new byte[] {0, 1}, new byte[] {}),
+          KV.of(new byte[] {1}, new byte[] {})
+        };
+
+    when(mockInMemorySorter.addIfRoom(kvs[0])).thenReturn(true);
+    when(mockInMemorySorter.addIfRoom(kvs[1])).thenReturn(true);
+    when(mockInMemorySorter.addIfRoom(kvs[2])).thenReturn(false);
+    when(mockInMemorySorter.sort()).thenReturn(Arrays.asList(kvs[0], kvs[1]));
+    when(mockExternalSorter.sort()).thenReturn(Arrays.asList(kvs[0], kvs[1], kvs[2]));
+
+    testSorter.add(kvs[0]);
+    testSorter.add(kvs[1]);
+    testSorter.add(kvs[2]);
+
+    Assert.assertEquals(Arrays.asList(kvs[0], kvs[1], kvs[2]), testSorter.sort());
+
+    verify(mockExternalSorter, times(1)).add(kvs[0]);
+    verify(mockExternalSorter, times(1)).add(kvs[1]);
+    verify(mockExternalSorter, times(1)).add(kvs[2]);
+  }
+
+  @Test
+  public void testEmpty() throws Exception {
+    SorterTestUtils.testEmpty(BufferedExternalSorter.create(new BufferedExternalSorter.Options()));
+  }
+
+  @Test
+  public void testSingleElement() throws Exception {
+    SorterTestUtils.testSingleElement(
+        BufferedExternalSorter.create(new BufferedExternalSorter.Options()));
+  }
+
+  @Test
+  public void testEmptyKeyValueElement() throws Exception {
+    SorterTestUtils.testEmptyKeyValueElement(
+        BufferedExternalSorter.create(new BufferedExternalSorter.Options()));
+  }
+
+  @Test
+  public void testMultipleIterations() throws Exception {
+    SorterTestUtils.testMultipleIterations(
+        BufferedExternalSorter.create(new BufferedExternalSorter.Options()));
+  }
+
+  @Test
+  public void testManySortersFewRecords() throws Exception {
+    SorterTestUtils.testRandom(
+        new SorterGenerator() {
+          @Override
+          public Sorter generateSorter() throws Exception {
+            return BufferedExternalSorter.create(new BufferedExternalSorter.Options());
+          }
+        },
+        1000000,
+        10);
+  }
+
+  @Test
+  public void testOneSorterManyRecords() throws Exception {
+    SorterTestUtils.testRandom(
+        new SorterGenerator() {
+          @Override
+          public Sorter generateSorter() throws Exception {
+            return BufferedExternalSorter.create(new BufferedExternalSorter.Options());
+          }
+        },
+        1,
+        1000000);
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void testAddAfterSort() throws Exception {
```

Comment:
please use the ExpectedException class and verify the expected error message and exception type. Can encapsulate this logic in the test class, if you like.

(and then just put `fail()` on line 2 of each test.